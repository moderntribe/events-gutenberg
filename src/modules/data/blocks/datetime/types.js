export const SET_START_DATE_TIME = 'SET_START_DATE_TIME';
export const SET_END_DATE_TIME = 'SET_END_DATE_TIME';

export const SET_DATES = 'SET_DATES';

export const SET_MULTI_DAY = 'SET_MULTI_DAY';
export const TOGGLE_MULTI_DAY = 'TOGGLE_MULTI_DAY';

export const SET_ALL_DAY = 'SET_ALL_DAY';

export const SET_SEPARATOR_DATE = 'SET_SEPARATOR_DATE';
export const SET_SEPARATOR_TIME = 'SET_SEPARATOR_TIME';

export const SET_TIME_ZONE = 'SET_TIME_ZONE';
