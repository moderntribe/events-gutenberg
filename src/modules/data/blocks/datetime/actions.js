/**
 * Internal dependencies
 */
import * as types from './types';

export const setStart = ( start ) => ( {
	type: types.SET_START_DATE_TIME,
	payload: {
		start,
	},
} );

export const setEnd = ( end ) => ( {
	type: types.SET_END_DATE_TIME,
	payload: {
		end,
	},
} );

export const setSeparatorDate = ( separator ) => ( {
	type: types.SET_SEPARATOR_DATE,
	payload: {
		separator,
	},
} );

export const setSeparatorTime = ( separator ) => ( {
	type: types.SET_SEPARATOR_TIME,
	payload: {
		separator,
	},
} );

export const setAllDay = ( allDay ) => ( {
	type: types.SET_ALL_DAY,
	payload: {
		allDay,
	},
} );

export const setMultiDay = ( multiDay ) => ( {
	type: types.SET_MULTI_DAY,
	payload: {
		multiDay,
	},
} );

export const setTimeZone = ( timezone ) => ( {
	type: types.SET_TIME_ZONE,
	payload: {
		timezone,
	},
} );
