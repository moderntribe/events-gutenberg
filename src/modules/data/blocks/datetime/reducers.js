/**
 * External dependencies
 */
import { combineReducers } from 'redux';

/**
 * Internal dependencies
 */
import values from './values/reducers';
import variants from './variants/reducers';
import separators from './separators/reducers';

export default combineReducers( {
	values,
	variants,
	separators,
} );
