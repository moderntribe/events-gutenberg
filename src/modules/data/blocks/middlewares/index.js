export { default as setInitialState } from './initial-state';
