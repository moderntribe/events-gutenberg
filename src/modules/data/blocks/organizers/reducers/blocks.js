import { combineReducers } from 'redux';

import * as types from './../types';
import block from './block';
const DEFAULT_STATE = {};

const byId = ( state = DEFAULT_STATE, action ) => {
	switch ( action.type ) {
		case types.ADD_ORGANIZER_BLOCK:
		case types.REMOVE_ORGANIZER_BLOCK:
			return {
				...state,
				[ action.payload.id ]: block( state[ action.payload.id ], action ),
			};
		default:
			return state;
	}
};

const allIds = ( state = [], action ) => {
	switch ( action.type ) {
		case types.ADD_ORGANIZER_BLOCK:
			return [ ...state, action.payload.organizer ];
		case types.REMOVE_ORGANIZER_BLOCK:
			return [ ...state ].filter( ( organizer ) => organizer !== action.payload.organizer );
		default:
			return state;
	}
};

export default combineReducers( {
	byId,
	allIds,
} );
