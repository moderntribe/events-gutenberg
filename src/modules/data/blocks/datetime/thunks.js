/**
 * WordPress dependencies
 */
import { applyFilters } from '@wordpress/hooks';

/**
 * Internal dependencies
 */
import {
	setStart,
	setEnd,
	setAllDay as setAllDayAction,
	setMultiDay as setMultiDayAction,
	setSeparatorDate,
	setSeparatorTime,
	setTimeZone,
} from './actions';
import { DEFAULT_STATE } from './reducers';
import { maybeBulkDispatch } from 'data/utils';
import {
	isSameDay,
	parseFormats,
	replaceDate,
	setTimeInSeconds,
	toDateTime,
	toMoment,
} from 'utils/moment';
import { HOUR_IN_SECONDS, DAY_IN_SECONDS } from 'utils/time';

export const setStartTime = ( { start, seconds } ) => ( dispatch ) => {
	const startDateTime = toDateTime( setTimeInSeconds( toMoment( start ), seconds ) );
	dispatch( setStart( startDateTime ) );
};

export const setEndTime = ( { end, seconds } ) => ( dispatch ) => {
	const endDateTime = toDateTime( setTimeInSeconds( toMoment( end ), seconds ) );
	dispatch( setEnd( endDateTime ) );
};

export const setAllDay = ( { start, end, isAllDay } ) => ( dispatch ) => {
	if ( isAllDay ) {
		const startDateTime = toDateTime( setTimeInSeconds( toMoment( start ), 0 ) );
		const endDateTime = toDateTime( setTimeInSeconds( toMoment( end ), DAY_IN_SECONDS - 1 ) );
		dispatch( setStart( startDateTime ) );
		dispatch( setEnd( endDateTime ) );
	}

	dispatch( setAllDayAction( isAllDay ) );
}

export const setDates = ( { start, end, from, to } ) => ( dispatch ) => {
	const startMoment = toMoment( start );
	const endMoment = toMoment( end );

	let newStartMoment = replaceDate( startMoment, toMoment( from ) );
	let newEndMoment = replaceDate( endMoment, toMoment( to || from ) );

	if ( newEndMoment.isSameOrBefore( newStartMoment ) ) {
		( { startMoment: newStartMoment, endMoment: newEndMoment } = resetTimes( newStartMoment ) );
	}

	dispatch( setStart( toDateTime( newStartMoment ) ) );
	dispatch( setEnd( toDateTime( newEndMoment ) ) );
};

export const setMultiDay = ( { start, end, isMultiDay } ) => ( dispatch ) => {
	if ( isMultiDay ) {
		const RANGE_DAYS = applyFilters( 'tec.datetime.defaultRange', 3 );
		const endMoment = toMoment( end ).clone().add( RANGE_DAYS, 'days' );
		dispatch( setEnd( toDateTime( endMoment ) ) );
	} else {
		let startMoment = toMoment( start );
		let endMoment = replaceDate( toMoment( end ), startMoment );

		if ( endMoment.isSameOrBefore( startMoment ) ) {
			( { startMoment, endMoment } = resetTimes( startMoment ) );
		}

		dispatch( setStart( toDateTime( startMoment ) ) );
		dispatch( setEnd( toDateTime( endMoment ) ) );
	}

	dispatch( setMultiDayAction( isMultiDay ) );
};

const resetTimes = ( startMoment ) => {
	const testMoment = startMoment.clone().add( HOUR_IN_SECONDS, 'seconds' );

	// Rollback half an hour before adding half an hour as we are on the edge of the day
	if ( ! isSameDay( startMoment, testMoment ) ) {
		startMoment.subtract( HOUR_IN_SECONDS, 'seconds' );
	}

	const endMoment = startMoment.clone().add( HOUR_IN_SECONDS, 'seconds' );

	return {
		startMoment,
		endMoment,
	};
};

export const setInitialState = ( { get, attributes } ) => ( dispatch ) => {
	const start = get( 'start', DEFAULT_STATE.start );
	const end = get( 'end', DEFAULT_STATE.end );

	maybeBulkDispatch( attributes, dispatch )( [
		[ setStart, 'start', DEFAULT_STATE.start ],
		[ setEnd, 'end', DEFAULT_STATE.end ],
		[ setAllDayAction, 'allDay', DEFAULT_STATE.allDay ],
		[ setSeparatorDate, 'separatorDate', DEFAULT_STATE.dateTimeSeparator ],
		[ setSeparatorTime, 'separatorTime', DEFAULT_STATE.timeRangeSeparator ],
		[ setTimeZone, 'timezone', DEFAULT_STATE.timezone ],
	] );

	if ( attributes.start ) {
		dispatch( setStart( toDateTime( parseFormats( attributes.start ) ) ) );
	}

	if ( attributes.end ) {
		dispatch( setEnd( toDateTime( parseFormats( attributes.end ) ) ) );
	}

	// sameDay
	const current = {
		start: toMoment( start ),
		end: toMoment( end ),
	};
	dispatch( setMultiDayAction( ! isSameDay( current.start, current.end ) ) );
};
