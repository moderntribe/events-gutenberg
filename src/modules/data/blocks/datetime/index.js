/**
 * Internal dependencies
 */
import reducer from './reducers';
import * as types from './types';
import * as actions from './actions';
import * as selectors from './selectors';

export default reducer;
export { types, actions, selectors };
