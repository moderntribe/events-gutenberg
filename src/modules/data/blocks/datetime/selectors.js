/**
 * External dependencies
 */
import { createSelector } from 'reselect';

/**
 * Internal dependencies
 */
import * as blockSelectors from 'data/blocks/selectors';

export const getAll = createSelector(
	[ blockSelectors.getBlocks ],
	( blocks ) => blocks.datetime,
);
