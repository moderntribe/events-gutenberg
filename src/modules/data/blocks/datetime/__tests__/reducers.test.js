/**
 * Internal dependencies
 */
import { actions } from 'data/blocks/datetime';
import reducer, { DEFAULT_STATE } from 'data/blocks/datetime/reducers';

jest.mock( 'moment', () => () => {
	const moment = require.requireActual( 'moment' );
	return moment( 'July 19, 2018 7:30 pm', 'MMMM D, Y h:mm a' );
} );

describe( '[STORE] - Datetime reducer', () => {
	it( 'Should set the default state', () => {
		expect( reducer( undefined, {} ) ).toEqual( DEFAULT_STATE );
	} );

	it( 'Should set the start', () => {
		expect( reducer( DEFAULT_STATE, actions.setStart( 'June 5, 2018 5:00 pm' ) ) )
			.toMatchSnapshot();
	} );

	it( 'Should set the end', () => {
		expect( reducer( DEFAULT_STATE, actions.setEnd( 'June 25, 2018 4:00 pm' ) ) ).toMatchSnapshot();
	} );

	it( 'Should set the start time', () => {
		expect( reducer( DEFAULT_STATE, actions.setStartTime( 72000 ) ) ).toMatchSnapshot();
	} );

	it( 'Should set the end time', () => {
		expect( reducer( DEFAULT_STATE, actions.setEndTime( 79200 ) ) ).toMatchSnapshot();
	} );

	it( 'Should set the separator time', () => {
		expect( reducer( DEFAULT_STATE, actions.setSeparatorTime( ' | ' ) ) ).toMatchSnapshot();
	} );

	it( 'Should set the separator date', () => {
		expect( reducer( DEFAULT_STATE, actions.setSeparatorDate( ' > ' ) ) ).toMatchSnapshot();
	} );

	it( 'Should set the timezone', () => {
		expect( reducer( DEFAULT_STATE, actions.setTimeZone( 'UTC' ) ) ).toMatchSnapshot();
	} );

	it( 'Should set the timezone label', () => {
		expect( reducer( DEFAULT_STATE, actions.setTimeZoneLabel( 'Modern Tribe' ) ) )
			.toMatchSnapshot();
	} );

	it( 'Should set the visibility of the timezone', () => {
		expect( reducer( DEFAULT_STATE, actions.setTimeZoneVisibility( true ) ) ).toMatchSnapshot();
		expect( reducer( DEFAULT_STATE, actions.setTimeZoneVisibility( false ) ) ).toMatchSnapshot();
	} );

	it( 'Should set the all day', () => {
		expect( reducer( DEFAULT_STATE, actions.setAllDay( true ) ) ).toMatchSnapshot();
		expect( reducer( DEFAULT_STATE, actions.setAllDay( false ) ) ).toMatchSnapshot();
	} );

	it( 'Should set the multi day', () => {
		expect( reducer( DEFAULT_STATE, actions.setMultiDay( true ) ) ).toMatchSnapshot();
		expect( reducer( DEFAULT_STATE, actions.setMultiDay( false ) ) ).toMatchSnapshot();
	} );

	it( 'Should toggle the multi day', () => {
		expect( reducer( DEFAULT_STATE, actions.toggleMultiDay() ) ).toMatchSnapshot();
	} );
} );
