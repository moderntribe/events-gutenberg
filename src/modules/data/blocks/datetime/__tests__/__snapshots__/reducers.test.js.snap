// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[STORE] - Datetime reducer Should set the all day 1`] = `
Object {
  "allDay": true,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the all day 2`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the end 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "June 25, 2018 4:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the end time 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the multi day 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": true,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the multi day 2`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the separator date 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": " > ",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the separator time 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": " | ",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the start 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "June 5, 2018 5:00 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the start time 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should set the timezone 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": false,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;

exports[`[STORE] - Datetime reducer Should toggle the multi day 1`] = `
Object {
  "allDay": false,
  "dateTimeSeparator": "@",
  "end": "July 19, 2018 8:00 pm",
  "multiDay": true,
  "start": "July 19, 2018 7:30 pm",
  "timeRangeSeparator": "-",
  "timezone": "UTC",
}
`;
